functions/src/download-readings.ts

```diff
@@ -0,0 +1,173 @@
+import * as functions from 'firebase-functions';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as os from 'os';
+import * as admin from 'firebase-admin';
+import SensorReading from './sensor-reading';
+import {firestore} from './admin';
+
+/**
+ * Uploads a file with name filename and specified data to the default storage
+ * bucket of the Firebase project.
+ * @param filename - name of file to be uploaded to Firebase bucket
+ * @param data - data to write to file
+ */
+function uploadFileToFirebaseBucket(filename: string, data: string) {
+  const tempLocalFile = path.join(os.tmpdir(), filename);
+  return new Promise<void>((resolve, reject) => {
+    // Write data into the temp file
+    fs.writeFile(tempLocalFile, data, error => {
+      if (error) {
+        reject(error);
+        return;
+      }
+
+      // Upload file into current Firebase project default bucket
+      admin
+        .storage()
+        .bucket()
+        .upload(tempLocalFile)
+        .then(() => resolve())
+        .catch(error => reject(error));
+    });
+  });
+}
+
+/**
+ * @param message - PubSub message that includes JSON with start and end fields
+ *
+ * @remarks
+ * The start and end values represent the non-inclusive interval of time to fetch
+ * readings for. The start and end values should be numbers that are the number
+ * of milliseconds since EPOCH.
+ *
+ * @example
+ * ```
+ * // Gets readings for the month of December (in PST)
+ * generateReadingsCsv(
+ *   {data: Buffer.from('{"start": 1606809599000, "end": 1609488000000}')}
```

Comment:
Is there a reason it has to use the raw epoch as opposed to something like a timestamp object or something that abstracts away the weird numbers?

Reply:
Since the date is in JSON, I thought that milliseconds since EPOCH would be the most consistent way to represent dates, especially since JavaScript recommends not using the constructors for dates that use strings.

This [StackOverflow post](https://stackoverflow.com/questions/10286204/what-is-the-right-json-date-format) sort of describes my thoughts, based on the final advice given: "The easiest and probably most portable format is the timestamp containing milliseconds since 1970"

Reply:
Yeah, definitely wouldn't want to do the string thing. I was just wondering if it would be possible to use the `FirebaseFirestore.Timestamp` type because I know that we'd been going back and forth to Firestore with that, but it's also possible I'm not understanding the use case here versus elsewhere.

Reply:
Parameters are passed to PubSub functions through JSON, so it's not possible to use a TypeScript class here

Reply:
That's annoying 😢